pkg/chunk/local/archiver.go

```diff
@@ -0,0 +1,435 @@
+
+	// We would use ingester name and startup timestamp for naming files while uploading so that
+	// ingester does not override old files when using same id
+	ingesterNameAndStartUpTs string
+}
+
+type file struct {
+	name  string
+	mtime time.Time
+	path  string
+}
+
+// NewArchiver creates an archiver for syncing local objects with a store
+func NewArchiver(cfg ArchiverConfig, boltDbDirectory string, archiveStoreClient ArchiveStoreClient) (*Archiver, error) {
+	if err := chunk_util.EnsureDirectory(cfg.CacheLocation); err != nil {
+		return nil, err
+	}
+
+	archiver := Archiver{
+		cfg:                      cfg,
+		boltDbDirectory:          boltDbDirectory,
+		downloadedTables:         map[string]*downloadedTable{},
+		updatesChan:              make(chan Update, updateChanCapacity),
+		uploadedFilesMtime:       map[string]time.Time{},
+		done:                     make(chan struct{}),
+		ingesterNameAndStartUpTs: fmt.Sprintf("%s-%d", cfg.IngesterName, time.Now().Unix()),
+	}
+
+	archiver.storeClient = archiveStoreClient
+
+	go archiver.loop()
+
+	return &archiver, nil
+}
+
+func (a *Archiver) loop() {
+	resyncTicker := time.NewTicker(a.cfg.ResyncInterval)
+	defer resyncTicker.Stop()
+
+	archiveFileTicker := time.NewTicker(ArchiverFileUploadInterval)
+	defer archiveFileTicker.Stop()
+
+	cacheTTLTicker := time.NewTicker(a.cfg.CacheTTL)
+	defer cacheTTLTicker.Stop()
+
+	for {
+		select {
+		case <-resyncTicker.C:
+			err := a.syncAllTables(context.Background())
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error syncing archived boltdb files with store", "err", err)
+			}
+		case <-archiveFileTicker.C:
+			err := a.archiveFiles(context.Background())
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error pushing archivable files to store", "err", err)
+			}
+		case <-cacheTTLTicker.C:
+			err := a.removeExpiredTables()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error cleaning up expired tables", "err", err)
+			}
+		case <-a.done:
+			return
+		}
+	}
+}
+
+// Stop the archiver and push all the local files to the store
+func (a *Archiver) Stop() {
+	close(a.done)
+
+	// Push all boltdb files to archive before returning
+	err := a.archiveFiles(context.Background())
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "error pushing archivable files to store", "err", err)
+	}
+}
+
+func (a *Archiver) archiveFiles(ctx context.Context) error {
+	if a.cfg.Mode == ArchiverModeReadOnly {
+		return nil
+	}
+
+	files, err := a.listFiles(a.boltDbDirectory)
+	if err != nil {
+		return err
+	}
+
+	a.uploadedFilesMtimeMtx.Lock()
+	defer a.uploadedFilesMtimeMtx.Unlock()
+
+	for _, file := range files {
+		// Checking whether file is updated after last push, if not skipping it
+		uploadedFileMtime, isOK := a.uploadedFilesMtime[file.path]
+		if isOK && !uploadedFileMtime.Before(file.mtime) {
+			continue
+		}
+
+		f, err := os.Open(file.path)
+		if err != nil {
+			return err
+		}
+
+		buf, err := ioutil.ReadAll(f)
```

Comment:
We're reading a file while BoltDB may simultaneously write on it. How can we guarantee the uploaded file is not corrupted?